optuna/samplers/_tpe/sampler.py

```diff
@@ -742,52 +736,34 @@ def _calculate_weights_below_for_multi_objective(
-
-    # Calculate weights based on hypervolume contributions.
-    n_below = len(lvals)
-    weights_below: np.ndarray
-    if n_below == 0:
-        weights_below = np.asarray([])
-    elif n_below == 1:
-        weights_below = np.asarray([1.0])
+    if constraints_func is None:
+        is_feasible = np.ones(len(below_trials), dtype=bool)
     else:
-        worst_point = np.max(lvals, axis=0)
-        reference_point = np.maximum(1.1 * worst_point, 0.9 * worst_point)
-        reference_point[reference_point == 0] = EPS
-        hv = compute_hypervolume(lvals, reference_point)
-        indices_mat = ~np.eye(n_below).astype(bool)
-        contributions = np.asarray(
-            [
-                hv - compute_hypervolume(lvals[indices_mat[i]], reference_point)
-                for i in range(n_below)
-            ]
-        )
-        contributions[np.isnan(contributions)] = np.inf
-        max_contribution = np.maximum(np.max(contributions), EPS)
-        if not np.isfinite(max_contribution):
-            weights_below = np.ones_like(contributions, dtype=float)
-            # TODO(nabenabe0928): Make the weights for non Pareto solutions to zero.
-            weights_below[np.isfinite(contributions)] = EPS
-        else:
-            weights_below = np.clip(contributions / max_contribution, EPS, 1)
+        is_feasible = np.asarray([all(c <= 0 for c in constraints_func(t)) for t in below_trials])
+
+    n_below_feasible = np.count_nonzero(is_feasible)
+    # For now, EPS weight is assigned to unpromising (e.g., infeasible) trials.
+    weights_below = np.full(len(below_trials), EPS)
+    if n_below_feasible == 0:
+        return weights_below
 
-    # For now, EPS weight is assigned to infeasible trials.
-    weights_below_all = np.full(len(below_trials), EPS)
-    weights_below_all[feasible_mask] = weights_below
-    return weights_below_all
+    lvals = np.asarray([t.values for t in below_trials])[is_feasible]
+    lvals *= np.array([-1.0 if d == StudyDirection.MAXIMIZE else 1.0 for d in study.directions])
+    ref_point = _get_reference_point(lvals)
+    on_front = _is_pareto_front(lvals, assume_unique_lexsorted=False)
+    hv = compute_hypervolume(lvals[on_front], ref_point, assume_pareto=True)
+    if np.isinf(hv) or n_below_feasible == 1:
+        weights_below[is_feasible] = 1.0
+        # TODO(nabenabe): Assign EPS to non-Pareto solutions, and
+        # solutions with finite contrib if hv is inf. Ref: PR#5813.
+        return weights_below
+
+    contribs = np.zeros(n_below_feasible, dtype=float)
+    # NOTE(nabenabe): Hypervolume contribution of a non-Pareto solution is zero.
+    pareto_indices = ~np.eye(n_below_feasible, dtype=bool)[on_front]
+    contribs[on_front] = hv - np.asarray(
+        [compute_hypervolume(lvals[indices], ref_point) for indices in pareto_indices]
+    )
+    # Calculate weights based on hypervolume contributions.
+    weights_below[is_feasible] = np.maximum(contribs / max(np.max(contribs), EPS), EPS)
```

Comment:
Please confirm this line.
The following might be the correct implementation:
```suggestion
    weights_below[is_feasible] = np.maximum(contribs / hv), EPS)
```

Reply:
Let $H(S, r)$ be a hypervolume of a set $S$ given a reference point $r$.
Then did you mean each weight $w$ should be $1 - \frac{H(S \setminus \\{s_i\\}, r)}{H(S, r)}$ rather than $\frac{H(S \setminus \\{s_i\\}, r)}{H(S, r) - \min_{i \in [N]} H(S \setminus \\{s_i\\}, r)}$?